### PR TITLE
Add tests for DSS on NVIDIA GPUs and only CPUs (New)

### PR DESCRIPTION
**Pull request by motjuste**
#### Changes to tests jobs

- Individual test jobs that used to check for available GPUs have been replaced by using the `graphics_card` resource, and enable skipping respective tests when the relevant GPUs are not available.
- Tests have been added to check setting up DSS for using CUDA with NVIDIA GPUs, and then some simple tests are run to see if PyTorch and Tensorflow can actually use the GPUs.
- Tests have been added to run only on the CPU. These tests will be run on all machines irrespective of available GPUs.
- Shell scripts have been refactored to have more re-usable functions.
  + This was partly done in lieu of investigating if these could have been implemented as Python scripts instead, but now they are far more compact than the Python scripts would have been.  Hence, re-implementation in Python has been dropped for now.
- Tests verifying Intel GPU were updated since they had a bug where they started counting NVIDIA GPUs too as Intel GPUs.  The tests are less precise now, i.e. now testing for a _minimum_ expected value for GPU counts and capacity to be available, instead of previous tests looking for exact counts.
    
#### Changes to the test plan

- Changes related to `resource.pxu` and additional NVIDIA tests, as explained above.
- Unused test-plans for individually testing ITEX or IPEX have been removed since they get tested in the main test plan any way.

#### Changes to the snap

- The command to trigger the tests from the `checkbox-dss` snap produced with the provider has been changed from `validate-intel-gpu` to `validate-with-gpu`.
- The command to install all the dependencies for the running the tests called `install-deps` has been refactored, and now accepts specifying version of the main snaps to be installed, which currently include DSS itself, Microk8s, and `kubectl`.
- These are backwards-incompatible change to the snap and hence its version has been bumped from `2.0` to `3.0`, and changes have been made to the relevant `snapcraft.yaml` and to the README.

#### Changes to the relevant GitHub workflow

- The GitHub workflow for running DSS has been refactored to now need a single job definition that can be used for all the values from the test matrix.
- An NVIDIA DGX machine has been added as a target machine, representing a machine that does not have any Intel GPUs, instead, only NVIDIA GPUs.
- Multiple Microk8s versions have been added to the test-matrix.

## Resolved issues

- [CHECKBOX-1586](https://warthogs.atlassian.net/browse/CHECKBOX-1586)
- [CHECKBOX-1668](https://warthogs.atlassian.net/browse/CHECKBOX-1668)

## Documentation

No changes to Checkbox's documentation.

## Tests

These DSS validations need to be run on machines from Testflinger.  See a recent run of the workflow [here](https://github.com/canonical/checkbox/actions/workflows/testflinger-contrib-dss-regression.yaml) (the relevant one for this PR is https://github.com/canonical/checkbox/actions/runs/12056842710). 

[CHECKBOX-1586]: https://warthogs.atlassian.net/browse/CHECKBOX-1586?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by motjuste** [comment]
Updated the PR with refactored scripts and decided against re-implementing them in Python.  Furthermore, since last week, I have further lumped in work for [CHECKBOX-1668](https://warthogs.atlassian.net/browse/CHECKBOX-1668) enabling customisation of Microk8s version in the `install-deps` script in order to lump together changes that require bumping changes to the version of the Snap.

File: contrib/checkbox-dss-validation/checkbox-provider-dss/bin/check_cuda [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        echo "Test success: NVIDIA GPU validations were successful!"
+    else
+        >&2 echo "Test failure: NVIDIA GPU validations were not successful, got ${result}"
+        exit 1
+    fi
+}
+
+help_function() {
+    echo "This script is used for tests related to CUDA"
+    echo "Usage: check_dss.sh <test_case>"
```
[the remark this answers is not in the document]
Comment (pieqq) [reply contrib/checkbox-dss-validation/checkbox-provider-dss/bin/check_cuda:39]: That said, I would refrain from removing the `.sh` extension. (see my general comment above)

**Review by pieqq — CHANGES_REQUESTED** [review]
Thanks for this big contribution! As usual, the very clear description and git commit messages help a lot with the review, along with the tests showing a successful run in TF.

Two things:

1. I would refrain from removing `.sh` extension to Shell scripts. It's much easier to see what kind of file it is by looking at the extension when the scripts are in the `bin/` directory
2. There is already a `graphics_card` resource in Checkbox that should help with checking if there is at least one Intel/NVIDIA GPU available in the system. Check my inline comment for more info on how to use it.

**Comment by motjuste** [comment]
The relevant workflow run in Testflinger for the latest commit accommodating the requested changes: https://github.com/canonical/checkbox/actions/runs/12122214006

**Comment by motjuste** [comment]
> The relevant workflow run in Testflinger for the latest commit accommodating the requested changes: https://github.com/canonical/checkbox/actions/runs/12122214006

You will see that all jobs testing latest/edge of DSS fail here unfortunately.  There was a release on this risk level yesterday for the DSS snap and it seems to have some bug (Issue reported [here](https://github.com/canonical/data-science-stack/issues/191)).  Since these are not failures of the validation suite, I believe then this PR is ready.

**Comment by motjuste** [comment]
The latest commit is a minor change to the README and does not impact the code, so I propose not to re-run the validations in Testflinger.

**Review by pieqq — APPROVED** [review]
Thanks for the modifications! LGTM :)

**Pull request merged**